### PR TITLE
Orders hw

### PR DESCRIPTION
**Pull request by Victoria-Moiseienko**

### CODE REVIEW

File: src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        }
+    }
+
+    @Override
+    public List<Order> getAllByUser(User user) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<Order> query = session.createQuery(
+                    "FROM Order o "
+                    + "JOIN FETCH o.tickets "
+                    + "JOIN FETCH o.user "
```
Comment (jah701) [inline src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java]: Check if you can make this query more simple and remove this redundant request

**Review by jah701 — APPROVED** [review]
Well done

File: src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java (hunk 1 above)
Comment (dariiabutorina) [inline src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java]: Do we need to fetch the user?
Which **_fetch_** type is set by default in the `@ManyToOne` and `@OneToOne` connections?

File: src/main/java/vic/cinema/service/impl/OrderServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        return orderDao.getAllByUser(user);
+    }
+
+    @Override
+    public Order completeOrder(List<Ticket> tickets, User user) {
+        Order order = new Order();
+        order.setTickets(tickets);
+        order.setUser(user);
+        order.setOrderDate(LocalDateTime.now());
+        return orderDao.create(order);
```
Comment (dariiabutorina) [inline src/main/java/vic/cinema/service/impl/OrderServiceImpl.java]: What should happen to the shopping cart's tickets after completing the order?

File: src/main/java/vic/cinema/Main.java [hunk 3]
```diff
@@ -76,8 +77,13 @@ public static void main(String[] args) throws AuthenticationException {
         shoppingCartService.addSession(movieSession, user2);
         System.out.println("Cart with tickets: ");
         System.out.println(shoppingCartService.getByUser(user2));
+
+        System.out.println("---- Order ----");
+        OrderService orderService =
+                (OrderService) injector.getInstance(OrderService.class);
+        orderService.completeOrder(shoppingCartService.getByUser(user2).getTickets(), user2);
         shoppingCartService.clear(shoppingCartService.getByUser(user2));
```
Comment (KseniiaMakarova) [inline src/main/java/vic/cinema/Main.java]: this should be done inside `completeOrder()`

File: src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java (hunk 1 above)
Comment (KseniiaMakarova) [reply src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java]: @DariiaPikul @jah701 let's test if the additional request is made to db to eager-fetch user
if yes, then it is nothing wrong with join fetching user

unfortunately, it seems that fetchType.EAGER does not guarantee that all associated entities will be fetched with one request
https://vladmihalcea.com/eager-fetching-is-a-code-smell/

File: src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                    + "WHERE o.user = :user ",
+                    Order.class);
+            query.setParameter("user", user);
+            return query.getResultList();
```
Comment (KseniiaMakarova) [inline src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java:51]: let's return only distinct values

File: src/main/java/vic/cinema/service/impl/OrderServiceImpl.java (hunk 2 above)
Comment (KseniiaMakarova) [reply src/main/java/vic/cinema/service/impl/OrderServiceImpl.java]: I agree

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java (hunk 1 above)
Comment (Victoria-Moiseienko) [reply src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java]: Fetch type for @OneToOne is EAGER. But without "JOIN FETCH o.user " I got 3 select-queries in a traice. In my version (with "JOIN FETCH o.user ") - I got 2 select-queries. 

For version with 1 select-query:
+ "JOIN FETCH t.movieSession ms "
+ "JOIN FETCH ms.cinemaHall "
+ "JOIN FETCH ms.movie "

File: src/main/java/vic/cinema/service/impl/OrderServiceImpl.java (hunk 2 above)
Comment (Victoria-Moiseienko) [reply src/main/java/vic/cinema/service/impl/OrderServiceImpl.java]: Done

File: src/main/java/vic/cinema/Main.java (hunk 3 above)
Comment (Victoria-Moiseienko) [reply src/main/java/vic/cinema/Main.java]: Done

File: src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java (hunk 4 above)
Comment (Victoria-Moiseienko) [reply src/main/java/vic/cinema/dao/impl/OrderDaoImpl.java:51]: Done

File: src/main/java/vic/cinema/service/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -6,26 +6,30 @@
...
         return orderDao.getAllByUser(user);
     }
 
     @Override
-    public Order completeOrder(List<Ticket> tickets, User user) {
+    public Order completeOrder(ShoppingCart cart) {
         Order order = new Order();
-        order.setTickets(tickets);
-        order.setUser(user);
+        order.setTickets(cart.getTickets());
```
Comment (KseniiaMakarova) [inline src/main/java/vic/cinema/service/impl/OrderServiceImpl.java]: it is better to set a copy of tickets list here
I addressed this issue yesterday in Slack

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**